File: compilers/go/compiler/n095_synthetic_array_chain_test.go
package compiler

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// ---------------------------------------------------------------------------
// N-095 — `syntheticArrayChain` on ANFProperty is wire data, and the wire is
// governed by packages/runar-ir-schema/src/schemas/anf-ir.schema.json.
//
// The expand-fixed-arrays pass desugars a `FixedArray` property into scalar
// siblings and hangs a chain of `{base, index, length}` levels off each leaf.
// The artifact assembler regroups those siblings back into a single FixedArray
// state/ABI entry by reading that chain off the *ANF program*, not off the AST
// -- so an ANF that loses the field still compiles to identical script bytes
// but degrades the SDK's `state.grid` accessor into N raw scalars. That is why
// the field must be BOTH emitted and declared in the shared schema: `$defs.
// ANFProperty` is `additionalProperties: false`, so an undeclared field makes
// Go's own ANF fail `validateANF`.
//
// Go is the tier whose spelling won (`syntheticArrayChain`); these tests pin
// it against the schema so a future rename in either place fails here.
// ---------------------------------------------------------------------------

const n095NestedFixedArraySource = `import { StatefulSmartContract, assert } from 'runar-lang';
import type { FixedArray } from 'runar-lang';

export class Grid2x2 extends StatefulSmartContract {
  grid: FixedArray<FixedArray<bigint, 2>, 2> = [[0n, 0n], [0n, 0n]];

  constructor() {
    super();
  }

  public set00(v: bigint) {
    this.grid[0][0] = v;
    assert(true);
  }

  public set11(v: bigint) {
    this.grid[1][1] = v;
    assert(true);
  }
}
`

// n095AnfSchemaProperties returns the key set `$defs.ANFProperty` accepts.
func n095AnfSchemaProperties(t *testing.T) map[string]bool {
	t.Helper()
	path := filepath.Join("..", "..", "..", "packages", "runar-ir-schema", "src", "schemas", "anf-ir.schema.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read ANF schema: %v", err)
	}
	var schema struct {
		Defs map[string]struct {
			AdditionalProperties *bool                      `json:"additionalProperties"`
			Properties           map[string]json.RawMessage `json:"properties"`
		} `json:"$defs"`
	}
	if err := json.Unmarshal(raw, &schema); err != nil {
		t.Fatalf("parse ANF schema: %v", err)
	}
	def, ok := schema.Defs["ANFProperty"]
	if !ok {
		t.Fatal("$defs.ANFProperty missing from the ANF schema")
	}
	if def.AdditionalProperties == nil || *def.AdditionalProperties {
		t.Fatal("$defs.ANFProperty is no longer additionalProperties:false — " +
			"this test's premise (an undeclared key is a schema violation) no longer holds")
	}
	keys := make(map[string]bool, len(def.Properties))
	for k := range def.Properties {
		keys[k] = true
	}
	return keys
}

// n095EmitIR compiles the source to ANF and returns the same JSON shape
// `--emit-ir` prints (marshalled through the ir package's json tags).
func n095EmitIR(t *testing.T, source string) []map[string]json.RawMessage {
	t.Helper()
	dir := t.TempDir()
	path := filepath.Join(dir, "Grid2x2.runar.ts")
	if err := os.WriteFile(path, []byte(source), 0o644); err != nil {
		t.Fatalf("write source: %v", err)
	}
	program, err := CompileSourceToIR(path)
	if err != nil {
		t.Fatalf("CompileSourceToIR: %v", err)
	}
	raw, err := json.Marshal(program)
	if err != nil {
		t.Fatalf("marshal ANF: %v", err)
	}
	var doc struct {
		Properties []map[string]json.RawMessage `json:"properties"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("unmarshal ANF: %v", err)
	}
	return doc.Properties
}

// TestN095_EmittedPropertyKeysAreAllDeclaredInTheSchema is the schema-
// conformance guard: every key Go writes on an ANFProperty must be a key the
// cross-tier schema declares. Before N-095 this failed on
// `syntheticArrayChain`.
func TestN095_EmittedPropertyKeysAreAllDeclaredInTheSchema(t *testing.T) {
	allowed := n095AnfSchemaProperties(t)
	for _, prop := range n095EmitIR(t, n095NestedFixedArraySource) {
		for key := range prop {
			if !allowed[key] {
				t.Errorf("emitted ANFProperty key %q is not declared in $defs.ANFProperty "+
					"(additionalProperties:false) — Go's own ANF fails validateANF", key)
			}
		}
	}
}

// TestN095_ExpandedLeavesCarryTheCamelCaseChain pins the spelling AND the
// shape. A tier that renames the key, or drops it, silently loses the ABI
// regrouping without moving a single script byte.
func TestN095_ExpandedLeavesCarryTheCamelCaseChain(t *testing.T) {
	props := n095EmitIR(t, n095NestedFixedArraySource)
	if len(props) != 4 {
		t.Fatalf("expected 4 expanded leaves, got %d", len(props))
	}
	want := [][2]int{{0, 0}, {0, 1}, {1, 0}, {1, 1}}
	for i, prop := range props {
		rawChain, ok := prop["syntheticArrayChain"]
		if !ok {
			t.Fatalf("leaf %d has no \"syntheticArrayChain\" key (keys: %v)", i, keysOf(prop))
		}
		if _, bad := prop["__syntheticArrayChain"]; bad {
			t.Errorf("leaf %d still carries the Rust AST-marker spelling", i)
		}
		if _, bad := prop["synthetic_array_chain"]; bad {
			t.Errorf("leaf %d still carries the Ruby snake spelling", i)
		}
		var chain []struct {
			Base   string `json:"base"`
			Index  int    `json:"index"`
			Length int    `json:"length"`
		}
		if err := json.Unmarshal(rawChain, &chain); err != nil {
			t.Fatalf("leaf %d chain does not match {base,index,length}: %v", i, err)
		}
		if len(chain) != 2 {
			t.Fatalf("leaf %d: chain length %d, want 2 (a 2x2 grid nests twice)", i, len(chain))
		}
		if chain[0].Base != "grid" || chain[0].Length != 2 || chain[0].Index != want[i][0] {
			t.Errorf("leaf %d outer level = %+v, want base=grid index=%d length=2", i, chain[0], want[i][0])
		}
		if chain[1].Length != 2 || chain[1].Index != want[i][1] {
			t.Errorf("leaf %d inner level = %+v, want index=%d length=2", i, chain[1], want[i][1])
		}
	}
}

// TestN095_ScalarPropertyCarriesNoChain is the byte-neutrality control: a
// contract with no FixedArray must not grow the key.
func TestN095_ScalarPropertyCarriesNoChain(t *testing.T) {
	const src = `import { StatefulSmartContract, assert } from 'runar-lang';

export class Counter extends StatefulSmartContract {
  count: bigint = 0n;

  constructor() {
    super();
  }

  public increment() {
    this.count = this.count + 1n;
    assert(true);
  }
}
`
	for i, prop := range n095EmitIR(t, src) {
		for key := range prop {
			if key == "syntheticArrayChain" {
				t.Errorf("property %d on a FixedArray-free contract grew a syntheticArrayChain key", i)
			}
		}
	}
}

// TestN095_SelfIRRoundTripStillRegroups is the acceptance test: Go's own ANF,
// fed back through `--ir`, must still produce the regrouped `grid` state field
// rather than four raw scalars.
func TestN095_SelfIRRoundTripStillRegroups(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "Grid2x2.runar.ts")
	if err := os.WriteFile(path, []byte(n095NestedFixedArraySource), 0o644); err != nil {
		t.Fatalf("write source: %v", err)
	}
	program, err := CompileSourceToIR(path)
	if err != nil {
		t.Fatalf("CompileSourceToIR: %v", err)
	}
	irJSON, err := json.Marshal(program)
	if err != nil {
		t.Fatalf("marshal ANF: %v", err)
	}

	fromSource, err := CompileFromSource(path)
	if err != nil {
		t.Fatalf("CompileFromSource: %v", err)
	}
	fromIR, err := CompileFromIRBytes(irJSON)
	if err != nil {
		t.Fatalf("CompileFromIRBytes: %v", err)
	}

	assertRegrouped(t, "source", fromSource)
	assertRegrouped(t, "ir", fromIR)

	if fromSource.Script != fromIR.Script {
		t.Errorf("script hex diverged between source mode and IR mode")
	}
}

func assertRegrouped(t *testing.T, label string, artifact *Artifact) {
	t.Helper()
	if len(artifact.StateFields) != 1 {
		names := make([]string, 0, len(artifact.StateFields))
		for _, sf := range artifact.StateFields {
			names = append(names, sf.Name)
		}
		t.Fatalf("%s: expected 1 regrouped state field, got %d %v", label, len(artifact.StateFields), names)
	}
	sf := artifact.StateFields[0]
	if sf.Name != "grid" {
		t.Errorf("%s: state field name = %q, want \"grid\"", label, sf.Name)
	}
	if sf.FixedArray == nil {
		t.Fatalf("%s: state field carries no fixedArray metadata — the SDK accessor is broken", label)
	}
	want := []string{"grid__0__0", "grid__0__1", "grid__1__0", "grid__1__1"}
	if len(sf.FixedArray.SyntheticNames) != len(want) {
		t.Fatalf("%s: syntheticNames = %v, want %v", label, sf.FixedArray.SyntheticNames, want)
	}
	for i, n := range want {
		if sf.FixedArray.SyntheticNames[i] != n {
			t.Errorf("%s: syntheticNames[%d] = %q, want %q", label, i, sf.FixedArray.SyntheticNames[i], n)
		}
	}
}

func keysOf(m map[string]json.RawMessage) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

File: compilers/java/src/test/java/runar/compiler/passes/N095SyntheticArrayChainTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.junit.jupiter.api.Test;
import runar.compiler.canonical.Jcs;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.AnfProperty;
import runar.compiler.ir.ast.ContractNode;

/**
 * N-095 — the synthetic-array chain on {@code AnfProperty} is wire data.
 *
 * <p>The expand-fixed-arrays pass desugars a {@code FixedArray} property into
 * scalar siblings and hangs a chain of {@code {base, index, length}} levels off
 * each leaf. Every other tier's artifact assembler regroups those siblings back
 * into a single FixedArray state/ABI entry by reading that chain off the ANF
 * PROGRAM, not off the AST. So an ANF that loses the field still compiles to
 * byte-identical script while the SDK's {@code state.grid} accessor degrades
 * into four raw {@code grid__i__j} scalars.
 *
 * <p>Java's {@code AnfProperty} carried no such component at all, so
 * {@code --emit-ir} dropped the chain entirely. That is a cross-tier defect
 * even though Java itself has no regrouper: an ANF Java emits is fed to every
 * other tier's {@code --ir}, and none of them could recover the regrouping
 * from it.
 *
 * <p><b>Deliberately out of scope here:</b> Java does not regroup
 * {@code stateFields} in EITHER mode — it has no port of the Go assembler's
 * {@code regroupStateFields}, as {@code Cli#writeArtifact}'s "Known gaps"
 * javadoc says. That is a separate, pre-existing defect. This test pins the
 * WIRE only: Java must write the chain and read it back under the settled
 * spelling, so the other six tiers can regroup Java's ANF.
 */
class N095SyntheticArrayChainTest {

    private static final String GRID_SRC = """
        import { StatefulSmartContract, assert } from 'runar-lang';
        import type { FixedArray } from 'runar-lang';

        export class Grid2x2 extends StatefulSmartContract {
          grid: FixedArray<FixedArray<bigint, 2>, 2> = [[0n, 0n], [0n, 0n]];

          constructor() {
            super();
          }

          public set00(v: bigint) {
            this.grid[0][0] = v;
            assert(true);
          }

          public set11(v: bigint) {
            this.grid[1][1] = v;
            assert(true);
          }
        }
        """;

    private static final String SCALAR_SRC = """
        import { StatefulSmartContract, assert } from 'runar-lang';

        export class Counter extends StatefulSmartContract {
          count: bigint = 0n;

          constructor() {
            super();
          }

          public increment() {
            this.count = this.count + 1n;
            assert(true);
          }
        }
        """;

    private static AnfProgram lower(String src, String fileName) throws Exception {
        ContractNode contract = ParserDispatch.parse(src, fileName);
        assertNotNull(contract, "parse produced no contract");
        return AnfLower.run(ExpandFixedArrays.run(contract));
    }

    /** The canonical ANF JSON {@code --emit-ir} prints. */
    private static String emitIr(String src, String fileName) throws Exception {
        return Jcs.stringify(lower(src, fileName));
    }

    // -----------------------------------------------------------------
    // Wire format
    // -----------------------------------------------------------------

    @Test
    void emitIrWritesTheChainUnderTheSettledSpelling() throws Exception {
        String ir = emitIr(GRID_SRC, "Grid2x2.runar.ts");

        assertTrue(ir.contains("\"syntheticArrayChain\""),
            "--emit-ir dropped the synthetic-array chain, so no other tier can regroup Java's ANF");
        assertFalse(ir.contains("\"__syntheticArrayChain\""), "AST-marker spelling on the wire");
        assertFalse(ir.contains("\"synthetic_array_chain\""), "snake spelling on the wire");

        // Four leaves, each with a two-level chain rooted at `grid`.
        assertEquals(4, countOccurrences(ir, "\"syntheticArrayChain\""));
        assertEquals(8, countOccurrences(ir, "\"base\""));
        assertEquals(4, countOccurrences(ir, "\"base\":\"grid\""));
    }

    @Test
    void everyEmittedPropertyKeyIsDeclaredInTheSharedSchema() throws Exception {
        Path schemaPath = Path.of("..", "..", "packages", "runar-ir-schema", "src", "schemas",
            "anf-ir.schema.json");
        String schema = Files.readString(schemaPath);

        // `$defs.ANFProperty` is additionalProperties:false, so anything Java
        // writes outside its declared key set fails validateANF.
        int defIdx = schema.indexOf("\"ANFProperty\"");
        assertTrue(defIdx >= 0, "$defs.ANFProperty missing from the shared ANF schema");
        String def = schema.substring(defIdx, schema.indexOf("\"ANFMethod\"", defIdx));
        assertTrue(def.contains("\"additionalProperties\": false"),
            "$defs.ANFProperty is no longer additionalProperties:false — this test's premise "
                + "(an undeclared key is a schema violation) no longer holds");

        for (String key : propertyKeys(emitIr(GRID_SRC, "Grid2x2.runar.ts"))) {
            assertTrue(def.contains('"' + key + '"'),
                "emitted ANFProperty key \"" + key + "\" is not declared in $defs.ANFProperty");
        }
    }

    /** Byte-neutrality control: a FixedArray-free contract must not grow the key. */
    @Test
    void scalarPropertyCarriesNoChain() throws Exception {
        String ir = emitIr(SCALAR_SRC, "Counter.runar.ts");
        assertFalse(ir.contains("ynthetic"),
            "a FixedArray-free contract grew a synthetic-array key: " + ir);
    }

    // -----------------------------------------------------------------
    // Loader readback — the half that makes Java a usable --ir consumer
    // -----------------------------------------------------------------

    @Test
    void loaderRecoversTheChainFromOurOwnAnf() throws Exception {
        AnfProgram loaded = AnfLoader.parse(emitIr(GRID_SRC, "Grid2x2.runar.ts"));

        assertEquals(4, loaded.properties().size());
        int[][] want = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
        for (int i = 0; i < loaded.properties().size(); i++) {
            AnfProperty p = loaded.properties().get(i);
            List<AnfProperty.SyntheticArrayLevel> chain = p.syntheticArrayChain();
            assertNotNull(chain, "leaf " + i + ": the loader dropped the chain");
            assertEquals(2, chain.size(), "leaf " + i + ": a 2x2 grid nests twice");
            assertEquals("grid", chain.get(0).base());
            assertEquals(want[i][0], chain.get(0).index());
            assertEquals(2, chain.get(0).length());
            assertEquals(want[i][1], chain.get(1).index());
            assertEquals(2, chain.get(1).length());
        }
    }

    /** Loading an ANF and re-emitting it must be a fixed point on the wire. */
    @Test
    void chainSurvivesLoadAndReEmit() throws Exception {
        String first = emitIr(GRID_SRC, "Grid2x2.runar.ts");
        String second = Jcs.stringify(AnfLoader.parse(first));
        assertEquals(first, second, "--emit-ir -> --ir -> --emit-ir is not a fixed point");
    }

    @Test
    void scalarPropertyLoadsWithNoChain() throws Exception {
        for (AnfProperty p : AnfLoader.parse(emitIr(SCALAR_SRC, "Counter.runar.ts")).properties()) {
            assertNull(p.syntheticArrayChain(), "a FixedArray-free property grew a chain on load");
        }
    }

    // -----------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------

    private static int countOccurrences(String haystack, String needle) {
        int n = 0;
        for (int i = haystack.indexOf(needle); i >= 0; i = haystack.indexOf(needle, i + 1)) n++;
        return n;
    }

    /** Every JSON key appearing inside the top-level {@code "properties"} array. */
    private static Set<String> propertyKeys(String ir) {
        int start = ir.indexOf("\"properties\":[");
        assertTrue(start >= 0, "no properties array in the emitted ANF");
        int depth = 0;
        int i = start + "\"properties\":".length();
        int end = i;
        for (; end < ir.length(); end++) {
            char c = ir.charAt(end);
            if (c == '[') depth++;
            else if (c == ']') {
                depth--;
                if (depth == 0) break;
            }
        }
        String slice = ir.substring(i, end + 1);
        Set<String> keys = new LinkedHashSet<>();
        Matcher m = Pattern.compile("\"([A-Za-z_][A-Za-z0-9_]*)\"\\s*:").matcher(slice);
        while (m.find()) keys.add(m.group(1));
        // The chain's own level keys are described by $defs.ANFSyntheticArrayLevel.
        keys.removeAll(Set.of("base", "index", "length"));
        return keys;
    }
}
